### PR TITLE
Added CosFace Loss

### PR DESCRIPTION
**Pull request by crsdvaibhav**
#### @monatis This is still work in progress. Just wanted to ask few questions, by sharing the code.
 - Is the implementation of the Multi Objective Loss according to your demands? If it is not, can you give a few pointers in the correct direction, and also about the tests?
 - CosFace Loss is correct according to me, as its only difference from Arcface Loss is the subtracting of margin from the cosines instead of angles. Could you give a pointer regarding writing the tests for the Cosface and Arcface losses?

Thankyou for the help!

### CODE REVIEW

**Comment by KarahanS** [comment]
Hi @crsdvaibhav, there is already a `l2_norm` function implemented for `arcface_loss`. If you are going to use `l2_norm` for `cos_face_loss` as well, then maybe it would be better to take `l2_norm` function to `loss/utils/utils.py` and call it from there in order to avoid code duplication.

File: quaterion/loss/cos_face_loss.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    """Apply L2 normalization to tensor
+    Args:
+        inputs: Input tensor.
+        dim: Dimension to operate on.
+    Returns:
+        torch.Tensor: L2-normalized tensor
+    """
+    outputs = inputs / torch.norm(inputs, 2, dim, True)
+
+    return outputs
```
Comment (KarahanS) [inline quaterion/loss/cos_face_loss.py:21]: Hi @crsdvaibhav, there is already a `l2_norm` function implemented for `arcface_loss`. If you are going to use `l2_norm` for `cos_face_loss` as well, then maybe it would be better to take `l2_norm` function to `loss/utils/utils.py` and call it from there in order to avoid code duplication.

File: quaterion/loss/multi_objective_loss.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
+from typing import Optional
+
+import torch
+
+
+class MultiObjectiveLoss(torch.nn.Module):
+    def __init__(self, losses: list, weights: Optional[list] = None):
```
Comment (KarahanS) [inline quaterion/loss/multi_objective_loss.py:7]: I think, in order to improve expressivity, that would be better to include data type stored in the list such as `weights: Optional[List[Tensor]]`.

**Comment by crsdvaibhav** [comment]
@KarahanS I have made the changes. Does this need some more changes before merge?

**Comment by KarahanS** [comment]
> @KarahanS I have made the changes. Does this need some more changes before merge?

Thanks for the updates! I'm not one of the official contributors of this project, so let's wait for the reviews by the maintainers for merging :)

**Comment by crsdvaibhav** [comment]
@monatis I have added the CosFace Loss and its test. Could you please review it? I have also added an implementation of Multi Objective Loss, could you please review it?

**Comment by monatis** [comment]
I'll review this PR, but could you please fix the conflict meanwhile? @crsdvaibhav

File: quaterion/loss/multi_objective_loss.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
+from typing import Optional
```
Comment (monatis) [inline quaterion/loss/multi_objective_loss.py:1]: This is a good start, but it has some issues:

- Not well documented with docstrings.
- Argument `groups` is missing in the `forward()` method.
- Tests are missing.

After fixing these issue it might be a good start, but I'm thinking of something slightly different that will make `GroupLoss` and `PairLoss` compatible with one another through some utility functions. For example, I got slightly better results when I combined Triplet Loss and Multiple Negatives Ranking Loss, but it will require more experiments.

tl; dr: Let's remove `MultiObjectiveLoss` from this PR for now --it will be something more ambitious and definitely is worth its own PR 😁

**Review by monatis — CHANGES_REQUESTED** [review]
- `CosFace` LGTM, but let's remove `MultiObjectiveLoss`.
- Fix conflicts.

File: quaterion/loss/multi_objective_loss.py (hunk 3 above)
Comment (crsdvaibhav) [reply quaterion/loss/multi_objective_loss.py:1]: Sure! I will update the PR with the CosFace Loss only. 

As for the issues with the Multi Objective Loss, I will surely solve them and put a new PR, and we can experiment with it there.

Thankyou!

Comment (crsdvaibhav) [reply quaterion/loss/multi_objective_loss.py]: I have done the required changes: 
- Removed the Multi Objecive Loss 
- Resolved the conflicts

**Comment by monatis** [comment]
I needed to close this. Could you please make another PR by rebasing from master and insuring that all the tests pass --because it didn't.
